### PR TITLE
Update Scanpy filter (new function)

### PR DESCRIPTION
**Pull request by heylf**
FOR CONTRIBUTOR:
* [X] I have read the [CONTRIBUTING.md](https://github.com/galaxyproject/tools-iuc/blob/master/CONTRIBUTING.md) document and this tool is appropriate for the tools-iuc repo.
* [X] License permits unrestricted use (educational + commercial)
* [ ] This PR adds a new tool or tool collection
* [X] This PR updates an existing tool or tool collection
* [ ] This PR does something else (explain below)

### CODE REVIEW

**Comment by heylf** [comment]
this adds to scanpy filter a function to filter marker genes (of a marker_gene.tsv tabular file) from an anndata object based on the mean expression and fraction of cells with that mean expression. 

This is useful for marker gene detection for spatial transcriptomics methods.

File: tools/scanpy/filter.xml [hunk 1]
```diff
@@ -98,6 +98,35 @@ sc.pp.downsample_counts(
...
+
+def check_marker(adata, group, gene, thresh_mean, thresh_frac, groupby):
+    filtered_data = adata[adata.obs[groupby] == group, adata.var_names == gene]
+    mean_expression = np.mean(filtered_data.X)
+    frac_cell_mean_expression = len(filtered_data.X[filtered_data.X > mean_expression]) / filtered_data.n_obs
+    if ( mean_expression > thresh_mean and frac_cell_mean_expression >= thresh_frac ):
+        return(True)
+    return(False)
+
+marker_list={key: list(value.values()) for key, value in pd.read_csv('$method.markerfile', sep='\t', index_col=0).to_dict(orient='index').items()}
```
Comment (pavanvidem) [inline tools/scanpy/filter.xml:116]: please consider keeping or discarding the first line based on whether it has markers or not.

File: tools/scanpy/filter.xml [hunk 2]
```diff
@@ -98,6 +98,35 @@ sc.pp.downsample_counts(
...
+
+#if $method.layer_selection.use_raw == 'False':
+    adata.X = adata.layers['$method.layer_selection.layer']
+#end if
+
+def check_marker(adata, group, gene, thresh_mean, thresh_frac, groupby):
+    filtered_data = adata[adata.obs[groupby] == group, adata.var_names == gene]
+    mean_expression = np.mean(filtered_data.X)
+    frac_cell_mean_expression = len(filtered_data.X[filtered_data.X > mean_expression]) / filtered_data.n_obs
+    if ( mean_expression > thresh_mean and frac_cell_mean_expression >= thresh_frac ):
```
Comment (pavanvidem) [inline tools/scanpy/filter.xml:112]: Is this a standard filtering approach for marker genes?

File: tools/scanpy/filter.xml [hunk 3]
```diff
@@ -444,6 +495,32 @@ sc.pp.downsample_counts(
...
+            </output>
+            <output name="anndata_out" file="cosg.rank_genes_groups.newton-cg.pbmc68k_highly_reduced_1.h5ad" ftype="h5ad">
+                <assert_contents>
+                    <has_h5_keys keys="obs, var, uns" />
+                </assert_contents>
+            </output>
+            <output name="marker_out" file="tl.rank_genes_groups.newton-cg.pbmc68k_highly_reduced_marker_filtered_1.tsv" ftype="tabular" compare="sim_size"/>
+        </test>
     </tests>
     <help><![CDATA[
```
Comment (pavanvidem) [inline tools/scanpy/filter.xml:526]: Please add a hint that users can generate a compatible marker genes tabular file using COSG tool

**Review by pavanvidem — COMMENTED** [review]
Looks great! Thanks @heylf

**Comment by heylf** [comment]
Review finished

**Review by pavanvidem — APPROVED** [review]
:+1: from my side

**Comment by mvdbeek** [comment]
Attention: deployment cancelled!

https://github.com/galaxyproject/tools-iuc/actions/runs/9616099750

**Pull request merged**